kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,96 @@
+package kyo
+
+import Layers.internal.*
+import java.util.concurrent.ConcurrentHashMap as JConcurrentHashMap
+import kyo.core.*
+import scala.annotation.targetName
+
+sealed trait Layer[+Out, -S]:
+    self =>
+
+    infix def to[Out2, S2, In2](that: Layer[Out2, Envs[In2] & S2]): Layer[Out2, S & S2]          = To(self, that)
+    infix def and[Out2, S2](that: Layer[Out2, S2]): Layer[Out & Out2, S & S2]                    = And(self, that)
+    infix def using[Out2, S2, In2](that: Layer[Out2, Envs[In2] & S2]): Layer[Out & Out2, S & S2] = self and (self to that)
+
+    private[kyo] def doRun(memoMap: JConcurrentHashMap[Layer[?, ?], Any] = JConcurrentHashMap()): TypeMap[Out] < (S & IOs) =
+        type Expected = TypeMap[Out] < (S & IOs)
+        memoMap.get(self) match
+            case nullable if isNull(nullable) =>
+                self match
+                    case And(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- rhs.doRun(memoMap)
+                            yield leftResult.union(rightResult)
+                        }.asInstanceOf[Expected]
+
+                    case To(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- Envs.runTypeMap(leftResult)(rhs.doRun(memoMap))(using summon, Envs.bypass)
+                            yield rightResult
+                        }.asInstanceOf[Expected]
+
+                    case FromKyo(kyo) =>
+                        kyo().map(result =>
+                            IOs {
+                                memoMap.putIfAbsent(self, result)
+                                result
+                            }
+                        ).asInstanceOf[Expected]
+
+            case result => result.asInstanceOf[Expected]
+        end match
+    end doRun
+
+end Layer
+
+object Layer:
+    import Envs.HasEnvs
+
+    extension [In, Out, S](layer: Layer[Out, Envs[In] & S])
+        def run[In1, R](using HasEnvs[In1, In] { type Remainder = R }): TypeMap[Out] < (S & R & IOs) =
+            layer.doRun().asInstanceOf
```

Comment:
do we need this method to be part of the public API? It feels like `Layers` is the proper API and `TypeMap` is more of a implementation detail. Also, please add the type annotation to `asInstanceOf`. It's possible the compiler ends up inferring `Nothing` in case we change this code.

Reply:
I went ahead and marked it private as I don't have a good example of using `.run` on a layer, but it may need to be public eventually.